src/pip/_internal/resolution/resolvelib/factory.py

```diff
@@ -104,32 +105,38 @@ def _make_candidate_from_link(
             return ExtrasCandidate(base, extras)
         return base
 
-    def make_candidate_from_ican(
-        self,
-        ican,  # type: InstallationCandidate
-        extras,  # type: Set[str]
-        parent,  # type: InstallRequirement
-    ):
-        # type: (...) -> Candidate
-        dist = self._installed_dists.get(ican.name)
-        should_use_installed_dist = (
-            not self._force_reinstall and
-            dist is not None and
-            dist.parsed_version == ican.version
+    def iter_found_candidates(self, ireq, extras):
+        # type: (InstallRequirement, Set[str]) -> Iterator[Candidate]
+        name = canonicalize_name(ireq.req.name)
+        if not self._force_reinstall:
+            dist = self._installed_dists.get(name)
+        else:
+            dist = None
+
+        found = self.finder.find_best_candidate(
+            project_name=ireq.req.name,
+            specifier=ireq.req.specifier,
+            hashes=ireq.hashes(trust_internet=False),
         )
-        if not should_use_installed_dist:
-            return self._make_candidate_from_link(
+        for ican in found.iter_applicable():
+            if dist is not None and dist.parsed_version == ican.version:
+                continue
+            yield self._make_candidate_from_link(
                 link=ican.link,
                 extras=extras,
-                parent=parent,
-                name=canonicalize_name(ican.name),
+                parent=ireq,
+                name=name,
                 version=ican.version,
             )
-        return self._make_candidate_from_dist(
-            dist=dist,
-            extras=extras,
-            parent=parent,
-        )
+
+        # Return installed distribution if it matches the specifier. This is
+        # done last so the resolver will prefer it over downloading links.
+        if dist is not None and dist.parsed_version in ireq.req.specifier:
+            yield self._make_candidate_from_dist(
+                dist=dist,
+                extras=extras,
+                parent=ireq,
+            )
```

Comment:
I think it would be useful to reverse the order of `find_matches` in resolvelib, specifically for this case, since we would be able to skip network I/O if the installed version is good enough.

Reply:
Or we can implement something like a “lazy list” that only evaluates when the elements are accessed with `__getitem__` or `__next__`. I’m going to open an issue for this as a possible optimisation.

Reply:
Recorded in #8023.